include/operations/NonUnitaryOperation.hpp

```diff
@@ -40,6 +40,18 @@ class NonUnitaryOperation final : public Operation {
   std::vector<Bit>& getClassics() { return classics; }
   [[nodiscard]] size_t getNclassics() const { return classics.size(); }
 
+  void addControls(const Controls& /*c*/) override {
+    throw QFRException("Cannot add controls to non-unitary operation.");
+  }
+
+  void clearControls() override {
+    throw QFRException("Cannot clear controls from non-unitary operation.");
+  }
+
+  void removeControls(const Controls& /*c*/) override {
+    throw QFRException("Cannot remove controls from non-unitary operation.");
+  }
```

Comment:
Would probably also make sense for consistency reasons to override the `getControls()` method to throw an exception here, right?

Reply:
I think yes, but this breaks one test (`CircuitDepthNonUnitaryOperations`, it calls `getControls` from the `Operation::addDepthContribution` function). I guess this can be fixed there by explicitly checking if an operation may have controls, but it might be cleaner to leave it as is. What do you think?

Reply:
I would rather argue that the proper way to fix this is to override the `addDepthContribution` method for the `NonUnitaryOperation` class so that it only ever iterates over the targets.
That should make the test work again, right?